src/Npgsql/NpgsqlConnection.cs

```diff
@@ -204,26 +205,23 @@ void SetupDataSource()
         // Note that we remove TargetSessionAttributes to make all connection strings that are otherwise identical point to the same pool.
         var canonical = settings.ConnectionStringForMultipleHosts;
 
-        if (PoolManager.Pools.TryGetValue(canonical, out _dataSource) && !_dataSource.NeedsRefresh())
+        lock (SetupDataSourceLock)
         {
-            // If this is a multi-host data source and the user specified a TargetSessionAttributes, create a wrapper in front of the
-            // MultiHostDataSource with that TargetSessionAttributes.
-            if (_dataSource is NpgsqlMultiHostDataSource multiHostDataSource && settings.TargetSessionAttributesParsed.HasValue)
-                _dataSource = multiHostDataSource.WithTargetSession(settings.TargetSessionAttributesParsed.Value);
-
-            // The pool was found, but only under the canonical key - we're using a different version
-            // for the first time. Map it via our own key for next time.
-            _dataSource = PoolManager.Pools.GetOrAdd(_connectionString, _dataSource);
-            return;
-        }
-
-        if (PoolManager.Pools.TryGetValue(canonical, out _dataSource) && _dataSource.NeedsRefresh())
-        {
-            _dataSource.Refresh();
-            // The pool was found, but only under the canonical key - we're using a different version
-            // for the first time. Map it via our own key for next time.
-            _dataSource = PoolManager.Pools.GetOrAdd(_connectionString, _dataSource);
-            return;
+            if (PoolManager.Pools.TryGetValue(canonical, out _dataSource))
+            {
+                if (_dataSource.NeedsRefresh())
+                {
+                    _dataSource.Refresh();
+                }
+                // If this is a multi-host data source and the user specified a TargetSessionAttributes, create a wrapper in front of the
+                // MultiHostDataSource with that TargetSessionAttributes.
+                if (_dataSource is NpgsqlMultiHostDataSource multiHostDataSource && settings.TargetSessionAttributesParsed.HasValue)
+                    _dataSource = multiHostDataSource.WithTargetSession(settings.TargetSessionAttributesParsed.Value);
+                // The pool was found, but only under the canonical key - we're using a different version
+                // for the first time. Map it via our own key for next time.
+                _dataSource = PoolManager.Pools.GetOrAdd(_connectionString, _dataSource);
```

Comment:
This line # 222 overrides the assignment to `_dataSource` done above in line # 219.
Why have the earlier assignment in the first place if it is getting overwritten immediately?

Reply:
This was part of the upstream code, it is not in any of our code paths since we reused Load Balance  Hosts property to redirect from the multihost policy to our cluster aware policy